### ISSUE TITLE
这个支持想simbertv2那样基于unilm同时进行bart式生成任务+对比学习语义理解任务训练吗

### ISSUE THREAD

**Issue by BigBen7** [JunnYu/RoFormer_pytorch#22]
大佬，请问下这个支持想simbertv2那样基于unilm同时进行bart式生成任务+对比学习语义理解任务训练吗？是能直接支持还是需要基于已经实现的接口组装一下，求指导，谢谢
https://kexue.fm/archives/8454

**Comment by JunnYu** [comment]
需要在已有代码基础上进行修改，添加Unified LM的Seq2Seq Mask 施加到 attention score上。

**Comment by JunnYu** [comment]
https://github.com/ZhuiyiTechnology/roformer-sim/blob/main/test/retrieval.py
以这个为例，
安装本仓库代码
将这个注释了`pooled_output = self.activation(pooled_output)`，
```python
class RoFormerPooler(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.dense = nn.Linear(config.hidden_size, config.hidden_size)
        self.activation = nn.Tanh()

    def forward(self, hidden_states):
        # We "pool" the model by simply taking the hidden state corresponding
        # to the first token.
        first_token_tensor = hidden_states[:, 0]
        pooled_output = self.dense(first_token_tensor)
        # pooled_output = self.activation(pooled_output)
        return pooled_output
```
因为 这代码使用的激活是线性，也就是不使用激活`with_pool='linear'`。
```python
bert = build_transformer_model(
    config_path,
    checkpoint_path,
    model='roformer',
    with_pool='linear',
    application='unilm',
    return_keras_model=False,
)
```
```python
from roformer import RoFormerModel,RoFormerTokenizer
import torch
import torch.nn as nn
model = RoFormerModel.from_pretrained('junnyu/roformer_chinese_sim_char_ft_base')
model.eval()
tokenizer = RoFormerTokenizer.from_pretrained('junnyu/roformer_chinese_sim_char_ft_base')

@torch.no_grad()
def similarity(text1, text2):
    """"计算text1与text2的相似度
    """
    inputs = tokenizer([text1,text2],padding=True, return_tensors='pt')
    Z = model(**inputs).pooler_output#.last_hidden_state

    Z /= (Z**2).sum(dim=1, keepdim=True)**0.5
    print((Z[0] * Z[1]).sum())


similarity('今天天气不错', '今天天气很好')
similarity('今天天气不错', '今天天气不好')
similarity('我喜欢北京', '我很喜欢北京')
similarity('我喜欢北京', '我不喜欢北京')
similarity('电影不错', '电影很好')
similarity('电影不错', '电影不好')
similarity('红色的苹果', '绿色的苹果')
similarity('给我推荐一款红色的车', '给我推荐一款黑色的车')
similarity('给我推荐一款红色的车', '推荐一辆红车')
similarity('给我推荐一款红色的车', '麻烦来一辆红车')
tensor(0.8944)
tensor(0.9218)
tensor(0.9760)
tensor(0.9028)
tensor(0.8880)
tensor(0.8944)
tensor(0.7801)
tensor(0.7981)
tensor(0.8914)
tensor(0.6807)

```

**Comment by JunnYu** [comment]
# 如果想要生成结果的话，下载这下面的roformer就可以生成结果了。
# [roformer.zip](https://github.com/JunnYu/RoFormer_pytorch/files/8402725/roformer.zip)

```python
import torch
import numpy as np
from roformer import RoFormerTokenizer, RoFormerForCausalLM, RoFormerConfig

device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
pretrained_model = "junnyu/roformer_chinese_sim_char_base"
tokenizer = RoFormerTokenizer.from_pretrained(pretrained_model)
config = RoFormerConfig.from_pretrained(pretrained_model)
config.is_decoder = True
config.eos_token_id = tokenizer.sep_token_id
config.pooler_activation = "linear"
model = RoFormerForCausalLM.from_pretrained(pretrained_model, config=config)
model.to(device)
model.eval()

def gen_synonyms(text, n=100, k=20):
    ''''含义： 产生sent的n个相似句，然后返回最相似的k个。
    做法：用seq2seq生成，并用encoder算相似度并排序。
    '''
    # 寻找所有相似的句子
    r = []
    inputs1 = tokenizer(text, return_tensors="pt")
    for _ in range(n):
        inputs1.to(device)
        output = tokenizer.batch_decode(model.generate(**inputs1, top_p=0.95, do_sample=True, max_length=128), skip_special_tokens=True)[0].replace(" ","").replace(text, "") # 去除空格，去除原始text文本。
        r.append(output)
    
    # 对相似的句子进行排序
    r = [i for i in set(r) if i != text and len(i) > 0]
    r = [text] + r
    inputs2 = tokenizer(r, padding=True, return_tensors="pt")
    with torch.no_grad():
        inputs2.to(device)
        outputs = model(**inputs2)
        Z = outputs.pooler_output.cpu().numpy()
    Z /= (Z**2).sum(axis=1, keepdims=True)**0.5
    argsort = np.dot(Z[1:], -Z[0]).argsort()
    
    return [r[i + 1] for i in argsort[:k]]

out = gen_synonyms("广州和深圳哪个好？")
print(out)
# ['深圳和广州哪个好？',
#  '广州和深圳哪个好',
#  '深圳和广州哪个好',
#  '深圳和广州哪个比较好。',
#  '深圳和广州哪个最好？',
#  '深圳和广州哪个比较好',
#  '广州和深圳那个比较好',
#  '深圳和广州哪个更好？',
#  '深圳与广州哪个好',
#  '深圳和广州，哪个比较好',
#  '广州与深圳比较哪个好',
#  '深圳和广州哪里比较好',
#  '深圳还是广州比较好？',
#  '广州和深圳哪个地方好一些？',
#  '广州好还是深圳好？',
#  '广州好还是深圳好呢？',
#  '广州与深圳哪个地方好点？',
#  '深圳好还是广州好',
#  '广州好还是深圳好',
#  '广州和深圳哪个城市好？']
```
